docs/RFCS/cluster_upgrade_tool.md

```diff
@@ -0,0 +1,107 @@
+before the cluster is upgraded to that version. This is expected to be
+infrequent; not every release will bump the version, only ones that need a
+migration hook.
+
+Migration hooks consist of a name, a work function, and optionally a minimum
+version. In the steady state, there is a map in code between migration names and
+the cockroach version they were released in. There is also a list of migrations
+added since the last cockroach release.
+
+When a node starts up, it checks that all known migrations have been run (the
+amount of work involved in this can be kept bounded using the `/SystemVersion`
+described below). If not, it runs them via a migration coordinator. (We should
+also do this check when a node rejoins a cluster it's been partitioned from, but
+I'm not sure how to do that.)
+
+The migration coordinator starts by heartbeating a kv entry to ensure that there
+is only ever one running at a time. Then for each missing migration, it runs the
+work function and writes a record of the completion to kv
+`/SystemVersion/<MigrationName>`. When finished, it writes its own version to
+`/SystemVersion`.
+
+The work function must be idempotent (in case a migration coordinator crashes)
+and runnable on some previous version of the cluster. If there is some minimum
+cluster version it needs, that is checked against the one written to
+`/SystemVersion`.
+
+The migration name to version map can be used to order migrations when a cluster
+needs to catch up many versions.
+
+We will introduce a thin cli wrapper `./cockroach cluster-upgrade` around
+starting the migration coordinator. When upgrading a Cockroach version, this
+command will be run and pointed at the cluster. After it returns, the cluster
+can be rolled onto the new version.
+
+The cli tool will be the recommended way of upgrading Cockroach versions for
+production scenarios. For ease of small clusters and local development, rolling
+one node will also do this upgrade before its normal startup. Any nodes rolled
+after the first, but before the cluster is upgraded, will panic to make it very
+clear the roll is unhealthy.
+
+When a migration is in progress, this should be exposed in the UI. It would be
+nice if we could also display the progress in the same place. Once we've used
+a migration to add a `system.jobs` table, it will be used for this.
+
+## Examples
+
+Simple migrations, like `CREATE TABLE IF NOT EXISTS system.jobs (...)`, can be
+accomplished with one hook and immediately used.
+
+Altering system tables will require more than one migration. For example, we'd
+like to remove the reference to `experimental_unique_bytes` from the
+`system.eventlog` table so we can delete the function. A migration is run to add
+a new column with a new name and the code in this Cockroach version sets both on
+mutations and fetches both (preferring the new one) on reads. Once the entire
+cluster is on this version, another code change stops reading and writing the
```

Comment:
Walk through the entire process of a multi-step update like this from the operator's perspective. Is it `cluster-upgrade` (add column), rollout binary version N (read and write both), `cluster-upgrade` (backfill new column), rollout binary version N+1 (stop using old column), `cluster-upgrade` (delete old column), rollout binary version N+2 (maybe not necessary)? 

That sounds terrible; we really don't want to put our users through a process like that. It's definitely not worth it to delete a tiny experimental function, although I assume we'll eventually have better reasons for making a similarly complex change. I think we need to be able to tolerate intermediate states so this can look like a single step (from the user's perspective). 

We don't need to solve this part now; it's more important that we get something that works for the simple cases. But this is part of why I still think it would be better to roll out the new binary before starting the migration process.